pkg/registration/hub/gc/controller.go

```diff
@@ -0,0 +1,110 @@
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	corev1informers "k8s.io/client-go/informers/core/v1"
+	rbacv1client "k8s.io/client-go/kubernetes/typed/rbac/v1"
+	rbacv1listers "k8s.io/client-go/listers/rbac/v1"
+	"k8s.io/client-go/tools/cache"
+
+	informerv1 "open-cluster-management.io/api/client/cluster/informers/externalversions/cluster/v1"
+	clusterv1listers "open-cluster-management.io/api/client/cluster/listers/cluster/v1"
+	worklister "open-cluster-management.io/api/client/work/listers/work/v1"
+	clusterv1 "open-cluster-management.io/api/cluster/v1"
+)
+
+type gcReconcileOp int
+
+const (
+	gcReconcileRequeue gcReconcileOp = iota
+	gcReconcileStop
+	gcReconcileContinue
+)
+
+// gcReconciler is an interface for reconcile cleanup logic after cluster is deleted.
+// clusterName is from the queueKey, cluster may be nil.
+type gcReconciler interface {
+	reconcile(ctx context.Context, clusterName string, cluster *clusterv1.ManagedCluster) (gcReconcileOp, error)
+}
+
+type GCController struct {
+	clusterLister clusterv1listers.ManagedClusterLister
+	gcReconcilers []gcReconciler
+}
+
+// NewGCController ensures the related resources are cleaned up after cluster is deleted
+func NewGCController(
+	clusterRoleLister rbacv1listers.ClusterRoleLister,
+	roleBindingLister rbacv1listers.RoleBindingLister,
+	namespaceInformer corev1informers.NamespaceInformer,
+	clusterInformer informerv1.ManagedClusterInformer,
+	manifestWorkLister worklister.ManifestWorkLister,
+	rbacClient rbacv1client.RbacV1Interface,
+	eventRecorder events.Recorder,
+) factory.Controller {
+	controller := &GCController{
+		clusterLister: clusterInformer.Lister(),
+		gcReconcilers: []gcReconciler{
+			newGCRoleBindingController(roleBindingLister, namespaceInformer, manifestWorkLister, rbacClient, eventRecorder),
+			newGCClusterRoleController(clusterRoleLister, clusterInformer, manifestWorkLister, rbacClient, eventRecorder),
+		},
+	}
+
+	// TODO: remove namespaceInformer and to prohibit delete cluster ns if the cluster is not deleting.
+	return factory.New().
+		WithInformersQueueKeysFunc(func(obj runtime.Object) []string {
+			accessor, _ := meta.Accessor(obj)
+			if accessor.GetDeletionTimestamp() != nil && !accessor.GetDeletionTimestamp().IsZero() {
+				return []string{accessor.GetName()}
+			}
+			return []string{}
+		}, clusterInformer.Informer(), namespaceInformer.Informer()).
+		WithSync(controller.sync).ToController("GCController", eventRecorder)
+}
+
+func (r *GCController) sync(ctx context.Context, controllerContext factory.SyncContext) error {
+	key := controllerContext.QueueKey()
+	if key == "" {
+		return nil
+	}
+
+	_, clusterName, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return nil
+	}
+
+	cluster, err := r.clusterLister.Get(clusterName)
+	if err != nil && !errors.IsNotFound(err) {
```

Comment:
should we remove finalizer on the cluster as the last step? if it is the gc controller of cluster.

Reply:
I think each finalizer should be removed by the the controller who adds it on the mcl in general. 
if the finalizers are not removed after a period of time for example 10 minutes, gc controller should remove all finalizers to force delete mcl.  
the force deletion strategy including the resources in cluster ns and mcl needs to be discussed more

Reply:
I do not think we should force remove the finalizer here. The only thing we can do is adding a condition stating why cluster is still deleting. Since the registration adds only one finalizer, I am also fine if we move adding finalizer to this controller.

Reply:
moved to this controller.